lib/ansible/modules/cloud/vmware/vmware_export_ovf.py

```diff
@@ -0,0 +1,328 @@
+    - '   folder: ha-datacenter/vm'
+    - '   folder: /datacenter1/vm'
+    - '   folder: datacenter1/vm'
+    - '   folder: /datacenter1/vm/folder1'
+    - '   folder: datacenter1/vm/folder1'
+    - '   folder: /folder1/datacenter1/vm'
+    - '   folder: folder1/datacenter1/vm'
+    - '   folder: /folder1/datacenter1/vm/folder2'
+  export_dir:
+    description:
+    - Absolute path to place the exported files on the server running this task, must have write permission.
+    - If folder not exist will create it, also create a folder under this path named with VM name.
+    required: yes
+  export_with_images:
+    default: false
+    description:
+    - Export an ISO image of the media mounted on the CD/DVD Drive within the virtual machine.
+    type: bool
+extends_documentation_fragment: vmware.documentation
+'''
+
+EXAMPLES = r'''
+- vmware_export_ovf:
+    validate_certs: false
+    hostname: '{{ vcenter_hostname }}'
+    username: '{{ vcenter_username }}'
+    password: '{{ vcenter_password }}'
+    name: '{{ vm_name }}'
+    export_with_images: true
+    export_dir: /path/to/ovf_template/
+  delegate_to: localhost
+'''
+
+RETURN = r'''
+instance:
+    description: list of the exported files, if exported from vCenter server, device file is not named with vm name
+    returned: always
+    type: dict
+    sample: None
+'''
+
+import os
+import hashlib
+from time import sleep
+from threading import Thread
+from ansible.module_utils.urls import open_url
+from ansible.module_utils.basic import AnsibleModule
+from ansible.module_utils._text import to_text
+from ansible.module_utils.vmware import (connect_to_api, vmware_argument_spec, PyVmomi)
+try:
+    from pyVmomi import vim
+    from pyVim import connect
+except ImportError:
+    pass
+
+
+class LeaseProgressUpdater(Thread):
+    def __init__(self, http_nfc_lease, update_interval):
+        Thread.__init__(self)
+        self._running = True
+        self.httpNfcLease = http_nfc_lease
+        self.updateInterval = update_interval
+        self.progressPercent = 0
+
+    def set_progress_percent(self, progress_percent):
+        self.progressPercent = progress_percent
+
+    def stop(self):
+        self._running = False
+
+    def run(self):
+        while self._running:
+            try:
+                if self.httpNfcLease.state == vim.HttpNfcLease.State.done:
+                    return
+                self.httpNfcLease.HttpNfcLeaseProgress(self.progressPercent)
+                sleep(self.updateInterval)
+            except Exception:
+                return
+
+
+class VMwareExportVmOvf(PyVmomi):
+    def __init__(self, module):
+        self.content = connect_to_api(module)
+        self.module = module
+        self.params = module.params
+        self.mf_file = ''
+        self.ovf_dir = ''
+        # set read device content chunk size to 2 MB
+        self.chunk_size = 2 * 2 ** 20
+        # set lease progress update interval to 30 seconds
+        self.lease_interval = 30
```

Comment:
30 seconds seems like a long time between progress updates?  It also guarantees up to 30 seconds of sleep time at the end of the operation as nothing interrupts the sleep before getting to done.

Reply:
Yeah, right, I made some changes to this, sleep every 1 second and check if the state is done, so this will not wait for 30 seconds when the lease is done. Thanks.